### ISSUE TITLE
Compiler error (illegal read) on tuple unpacking in a `for` loop

### ISSUE THREAD

**Issue by ZoomRmc** [nim-lang/Nim#23180]
### Description

```nim
iterator it(): bool =
  discard

for (_) in it():
  discard
```

The culprit is tuple unpacking in the for loop.

Had some fun minimizing the reproducing code: first I thought it's a proc-local iterator, then accessing variables outside of the iterator's scope, then returning a tuple, turns out it's the unpacking.

Also irrelevant: Usage of the underscore. Having an actual tuple for the iterator's return type.

### Nim Version

> Nim Compiler Version 2.1.1 [Linux: amd64]
> Compiled at 2023-12-21

Earliest I tried is 1.4.6

### Current Output

```text
> SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by juancarlospaco** [comment]
!nim c

```nim
iterator it(): bool = discard
for (_) in it(): discard
```

**Comment by metagn** [comment]
Putting parentheses makes it unpack tuples in for loops (old but broken and undocumented behavior, there are other issues open because of this: #19805, #15924). ~~Since `bool` is not a tuple type, the compiler crashes (I surmise). So we need to check that the type is a tuple.~~ (As written below, this is not the cause, but the issue does still seem to be with the parentheses) The for loop tuple unpacking code probably needs cleaning up, it wasn't touched during the recent tuple unpacking refactors

**Comment by ringabout** [comment]
```nim
iterator it(): bool =
  discard

for (x, ) in it():
  discard
```
It crashed like the single variable case. I suppose for loop tuple unpacking didn't consider this case

**Comment by ZoomRmc** [comment]
> Putting parentheses makes it unpack tuples in for loops (old but broken and undocumented behavior, there are other issues open because of this: #19805, #15924). Since `bool` is not a tuple type, the compiler crashes (I surmise). So we need to check that the type is a tuple. The for loop tuple unpacking code probably needs cleaning up, it wasn't touched during the recent tuple unpacking refactors

As I've mentioned in the issue, it doesn't matter if the returned type is a tuple or not, it still crashes. I just minimized the code as much as I possibly could.

```nim
proc main() =
  let x = false
  iterator it(): (bool, bool) =
    if x: discard
    discard

  for (_, _) in it():
    discard

main()
```

On the second thought, to crash with the actual tuple we need to access a captured local variable. So it's might be a separate bug :)

**Comment by metagn** [comment]
> As I've mentioned in the issue, it doesn't matter if the returned type is a tuple or not

My bad, didn't see.

There is a separate bug simplified from that example on 2.0 independent from tuple unpacking (probably same as #21476, same error, same workaround):

```nim
proc main =
  let x = false
  iterator it(): bool =
    discard x

  for a in it():
    discard
main()
```

which is obtained by removing the parentheses from that example, so the problem does still seem to be with the parentheses.

**Comment by ringabout** [comment]
I think the first example is invalid code, just as 
```nim
let (a) = false
```
gives 
```
Error: tuple expected for tuple unpacking, but got 'int'
```

**Comment by metagn** [comment]
There is still a bug like #21476 for parentheses but doesn't work with the workaround for #21476. Maybe it is still the same issue though.

```nim
proc main() =
  let x = false
  iterator it(): (bool, bool) {.closure.} = # normally {.closure.} here makes #21476 work
    discard x

  for (_, _) in it():
    discard

main()
```

Output is segfault.

**Comment by ringabout** [comment]
> Putting parentheses makes it unpack tuples in for loops (old but broken and undocumented behavior, there are other issues open because of this:

It's quite weird that the tuple unpacking works for nested arrays in the for loop. I didn't know this feature/bug before. It is quite inconsistent with var tuples unpacking.

**Comment by metagn** [comment]
I think the purpose is to prevent an implicit `pairs` iterator while allowing `items` but we need to make sure this is the only difference and document it

**Comment by ringabout** [comment]
I checked https://github.com/nim-lang/Nim/issues/19805. It seems to me that it happened to compile because there is no restriction on the tuple types for tuple unpacking. Though I will wait for https://github.com/nim-lang/Nim/pull/23185 to see whether it breaks something.

`errWrongNumberOfVariables` is cute though since it doesn't work for array types (`iterAfterVarLent.len` is always 2), it's reasonable to suppose there would have been a check for tuple types.
```nim
if n[0].len-1 != iterAfterVarLent.len:
  return localErrorNode(c, n, n[0].info, errWrongNumberOfVariables)

```
I think it's most likely unintentional. The cases in the https://github.com/nim-lang/Nim/issues/19805 failed twice to get itself compiled.

**Comment by ringabout** [comment]
I think the example two is the same case as https://github.com/nim-lang/Nim/issues/15924 because there is no var tuple unpacking for closures

**Comment by ringabout** [comment]
> There is still a bug like #21476 for parentheses but doesn't work with the workaround for #21476. Maybe it is still the same issue though.
> 
> ```nim
> proc main() =
>   let x = false
>   iterator it(): (bool, bool) {.closure.} = # normally {.closure.} here makes #21476 work
>     discard x
> 
>   for (_, _) in it():
>     discard
> 
> main()
> ```
> 
> Output is segfault.

fixed by https://github.com/nim-lang/Nim/pull/23205

**Comment by ZoomRmc** [comment]
> fixed by #23205

```nim
proc main() =
  let x = false
  iterator it(): (bool, bool) =
    if x: discard
    discard
  for (_, _) in it(): discard

main()
```
`> Error: internal error: proc has no result symbol`

**Comment by metagn** [comment]
That should be the same issue as #21476, as long as it doesn't matter if we use `(_, _)` or just `_`. Still worth mentioning in #21476 if it is the same issue though since the original example in #21476 is a bit misleading.
